File: tests/cpp_test_scenarios/src/test_basic.cpp
#include "test_basic.hpp"

#include <cassert>
#include <iostream>
#include <kvs.hpp>
#include <kvsbuilder.hpp>
#include <unordered_map>

#include "score/json/json_parser.h"
#include "score/json/json_writer.h"
#include "score/result/result.h"
#include "tracing.hpp"

namespace {

struct KvsParameters {
    uint64_t instance_id;
    std::optional<bool> need_defaults;
    std::optional<bool> need_kvs;
    std::optional<std::string> dir;
    std::optional<bool> flush_on_exit;
};

KvsParameters map_to_params(const std::string& data) {
    using namespace score::json;

    JsonParser parser;
    auto any_res{parser.FromBuffer(data)};
    if (!any_res) {
        throw std::runtime_error{"Failed to parse JSON data"};
    }
    const auto& map_root{any_res.value().As<Object>().value().get().at("kvs_parameters")};
    const auto& obj_root{map_root.As<Object>().value().get()};

    KvsParameters params;
    params.instance_id = obj_root.at("instance_id").As<double>().value();
    if (obj_root.find("need_defaults") != obj_root.end()) {
        params.need_defaults = obj_root.at("need_defaults").As<bool>().value();
    }
    if (obj_root.find("need_kvs") != obj_root.end()) {
        params.need_kvs = obj_root.at("need_kvs").As<bool>().value();
    }
    if (obj_root.find("dir") != obj_root.end()) {
        params.dir = obj_root.at("dir").As<std::string>().value();
    }
    if (obj_root.find("flush_on_exit") != obj_root.end()) {
        params.flush_on_exit = obj_root.at("flush_on_exit").As<bool>().value();
    }

    return params;
}

const std::string kTargetName{"cpp_test_scenarios::basic::basic"};

}  // namespace

std::string BasicScenario::name() const { return "basic"; }

void BasicScenario::run(const std::optional<std::string>& input) const {
    using namespace score::mw::per::kvs;

    // Print and parse parameters.
    std::cerr << *input << std::endl;

    auto params{map_to_params(*input)};

    // Set builder parameters.
    InstanceId instance_id{params.instance_id};
    KvsBuilder builder{instance_id};
    if (params.need_defaults.has_value()) {
        builder = builder.need_defaults_flag(*params.need_defaults);
    }
    if (params.need_kvs.has_value()) {
        builder = builder.need_kvs_flag(*params.need_kvs);
    }
    // TODO: handle dir?

    // Create KVS.
    Kvs kvs{*builder.build()};
    if (params.flush_on_exit.has_value()) {
        kvs.set_flush_on_exit(*params.flush_on_exit);
    }

    // Simple set/get.
    std::string key{"example_key"};
    std::string value{"example_value"};
    auto set_value_result{kvs.set_value(key, KvsValue{value})};
    if (!set_value_result) {
        throw std::runtime_error("Failed to set value");
    }

    auto get_value_result{kvs.get_value(key)};
    if (!get_value_result) {
        throw std::runtime_error{"Failed to get value"};
    }
    auto stored_kvs_value{get_value_result.value()};
    if (stored_kvs_value.getType() != KvsValue::Type::String) {
        throw std::runtime_error{"Invalid value type"};
    }

    auto stored_value{std::get<std::string>(stored_kvs_value.getValue())};
    if (stored_value.compare(value) != 0) {
        throw std::runtime_error("Value mismatch");
    }

    // Trace.
    TRACING_INFO(kTargetName, std::pair{std::string{"example_key"}, stored_value});
}

File: tests/rust_test_scenarios/src/test_basic.rs
use crate::helpers::kvs_parameters::KvsParameters;
use rust_kvs::prelude::*;
use test_scenarios_rust::scenario::Scenario;
use tracing::info;

fn _error_code_to_string(e: ErrorCode) -> String {
    format!("ErrorCode::{e:?}")
}

pub struct BasicScenario;

/// Checks (almost) empty program with only shutdown
impl Scenario for BasicScenario {
    fn name(&self) -> &'static str {
        "basic"
    }

    fn run(&self, input: Option<String>) -> Result<(), String> {
        // Print and parse parameters.
        let input_string = input.as_ref().expect("Test input is expected");
        eprintln!("{input_string}");

        let params = KvsParameters::from_json(input_string).expect("Failed to parse parameters");

        // Set builder parameters.
        let mut builder = KvsBuilder::new(params.instance_id);
        if let Some(flag) = params.defaults {
            builder = builder.defaults(flag);
        }
        if let Some(flag) = params.kvs_load {
            builder = builder.kvs_load(flag);
        }
        if let Some(dir) = params.dir {
            builder = builder.dir(dir.to_string_lossy().to_string());
        }

        // Create KVS.
        let kvs: Kvs = builder.build().expect("Failed to build KVS instance");

        // Simple set/get.
        let key = "example_key";
        let value = "example_value".to_string();
        kvs.set_value(key, value).expect("Failed to set value");
        let value_read = kvs
            .get_value_as::<String>(key)
            .expect("Failed to read value");

        // Trace.
        info!(example_key = value_read);

        Ok(())
    }
}
